### ISSUE TITLE
Adding Kappa into Evaluation Metric

### ISSUE THREAD

**Issue by hxd1011** [dmlc/xgboost#1152]
In R, I can use Cohen's Kappa as customized evaluation metric. But this will slow down the training process, because R is very slow. Do we want to have a C implementation of Kappa? If we want please point to the location I can contribute. xgboost/src/metric/ ?

``` R
kappa <- function(preds, y) {
  label=getinfo(y,"label")
  k=vcd::Kappa(table(preds>0.5,label))
  return(list(metric = "kappa", value = as.numeric(k$Unweighted[1])))
}
param <- list(  objective = "binary:logistic",
                booster = "gbtree",
                eval_metric = kappa)
```

**Comment by khotilov** [comment]
You can take the customized objective example for a start: https://github.com/dmlc/xgboost/tree/master/plugin/example

BTW, this function takes about 50-60 milliseconds per call with 100K cases. Which is about a minute for 1K iterations. The majority of time is spent calculating the contingency table.

**Comment by hxd1011** [comment]
Thanks for your reply khotilov, I will take a look to see if I can contribute.

You are exactly right, the function I wrote is extremely slow for a large data, that is why I asked this question at first place. I think the existing metric like "auc" calculation should be more complicated than this but really fast, and can run in parallel. could you tell me why "auc" would work as it is? is that because it was written in C?

**Comment by Laurae2** [comment]
For AUC, it is pretty fast and scales very well.

```
FastROC <- function(y, x) {
  
  # y = actual
  # x = predicted
  x1 = x[y==1]
  n1 = length(x1)
  x2 = x[y==0]
  n2 = length(x2)
  r = rank(c(x1,x2))
  return((sum(r[1:n1]) - n1*(n1+1)/2) / (n1*n2))
  
}
```

The implementation used in xgboost in using C. You can find it here: https://github.com/dmlc/xgboost/blob/master/src/metric/rank_metric.cc

Did you try vectorizing the counts?:

* A ==> count of label = 0 and prediction < 0.5
* B ==> count of label = 1 and prediction < 0.5
* C ==> count of label = 0 and prediction >= 0.5
* D ==> count of label = 1 and prediction >= 0.5
* N ==> A+B+C+D
* pObs ==> (A+D) / N
* pExp ==> ( (A+C)/N * (A+B)/N ) + ( (B+D)/N * (C+D)/N )
* Kappa ==> (pObs - pExp) / (1 - pExp)

Something like this should nail it:

```
...
# do the typical pre-processing, i.e: preds to predictions (because you are given the inputs before logistic regression), labels to variable via getinfo
A <- sum((label == 0) & (preds < 0.5))
B <- sum((label == 1) & (preds < 0.5))
C <- sum((label == 0) & (preds >= 0.5))
D <- sum((label == 1) & (preds >= 0.5))
N <- A+B+C+D
pObs <- (A+D)/N
pExp <- ((A+C)/N * (A+B)/N) + ((B+D)/N * (C+D)/N)
return(list(metric = "kappa", value = (pObs - pExp) / (1 - pExp)))
```

**Comment by hxd1011** [comment]
Hi, Laurae2 thank you very much for reply. I did use vectored counts "table(preds>0.5,label)" is the key part and runs extremely slow.

May be it is a R implementation or C implementation issue?

**Comment by Laurae2** [comment]
hxd1011: don't use table, it does a lot of hidden checkings with a loop. Use directly the vectorized computation without table. My version of Kappa (that was at the end of my previous post) should be much faster than your version.

For 100000 values, my vectorized function is so fast it cannot catch centiseconds.:

```
> system.time(Kappa(table(preds>0.5, labels))$Unweighted[1])
   user  system elapsed 
   0.11    0.00    0.11 

> #remember to add the raw->logistic predictions transformation
> my_func <- function(preds, label) { A <- sum((label == 0) & (preds < 0.5))
+ B <- sum((label == 1) & (preds < 0.5))
+ C <- sum((label == 0) & (preds >= 0.5))
+ D <- sum((label == 1) & (preds >= 0.5))
+ N <- A+B+C+D
+ pObs <- (A+D)/N
+ pExp <- ((A+C)/N * (A+B)/N) + ((B+D)/N * (C+D)/N)
+ list(metric = "kappa", value = (pObs - pExp) / (1 - pExp)) }
> system.time(my_func(preds, labels))
   user  system elapsed 
      0       0       0 

> my_func(preds, labels) #identical?
$metric
[1] "kappa"

$value
[1] 0.001479188

> Kappa(table(preds>0.5, labels))$Unweighted[1] #identical?
      value 
0.001479188 
```

**Comment by hxd1011** [comment]
Thank you very much Laurae2. 

You reminded me what I learned from this post.

http://www.r-bloggers.com/faster-higher-stonger-a-guide-to-speeding-up-r-code-for-busy-people/

"Standard and internal functions: A few standard functions in R are pretty slow, often because they perform error checking and methods dispatch before actually doing arithmetic. mean is among the most notorious of these. mean(x)does a lot more than just sum(x)/length(x). It checks the form of the inputs, assigns the appropriate method, then dispatches .Internal(mean(x)). The latter is much faster, as it is written in C"

I will try your code and try to make it into xgboost R library.

**Comment by khotilov** [comment]
@Laurae2 : the manual contingency table calculation would not generalize well to multiclass setting. I'm not a big fan of Kappa, but I have some interest in it for the multiclass use.

@hxd1011 : calling things "extremely slow" without providing any specific supporting numbers of your own is not productive here.

Please see the following gist where I compared a few contingency table calculation approaches: https://gist.github.com/khotilov/a0bd2f617e11811ede89287e42796ba5 
The tabulate-based method with predefined number of classes is about 50 times faster than table, works easily for multiclass, and is comparable to the manual calculation in a binary case. `table` is just a fancy safety and convenience wrapper around `tabulate`. Coding the method in C++ with multithreading might make it even faster for large samples, but I don't think that would provide much or any additional practical advantage.

Here is a modified fast version of the kappa metric for binary classification from the initial message:
```R
kappa <- function(preds, y) {
  label <- getinfo(y, "label")
  k <- vcd::Kappa(matrix(tabulate(1 + (preds>0.5) + 2*label, 4), ncol=2))
  return(list(metric = "kappa", value = as.numeric(k$Unweighted[1])))
}
```
However, to make it reusable, it would be better to wrap it into a closure which would allow to customize the function for multiclass setting, for thresholds different than 0.5, and also for using alternative definitions of Kappa. See the `create.xtab.nclass` function in my gist for a simple example of a closure. Also keep in mind that there are two multiclass objectives in xgboost, 'multi:softmax' and 'multi:softprob', producing discrete and probability predictions in different formats.

If the developers would agree, it might make sense to start a collection of some reusable custom objective and evaluation functions within the R-package.
